lib/submit-request.js

```diff
@@ -40,9 +41,51 @@ function SubmitRequest(backend, agent, index, id, op, options) {
   this.snapshot = null;
   this.ops = [];
   this.channels = null;
+  this._fixupOps = [];
 }
 module.exports = SubmitRequest;
 
+SubmitRequest.prototype.$fixup = function(op) {
+  if (this.action !== this.backend.MIDDLEWARE_ACTIONS.apply) {
```

Comment:
I'm not sure if this restriction is strictly necessary — we could theoretically allow consumers to fixup in the `commit` hook (say), and just `apply()` on the snapshot, but it felt like keeping the scope of this change narrow to start was better, because I'm not 100% sure I understand the ramifications of allowing that sort of behaviour.